KXI-59279 extending compression performance (#353)
KxSystems/docs: docs/kb/file-compression.md

## Performance

There are three key aspects of compression algorithms:

   1. **Compression ratio**: This indicates how much the final data file size is reduced. A high compression ratio means smaller files and lower storage, I/O costs. If the column files are smaller, we can store more data on a storage of a given size. Similarly, more storage space costs more (especially in the cloud). Smaller files may reduce query execution time if the storage is slow because smaller files are read.

>    1. **Compression and encryption speed**: This measures the time required to compress a file. Compression is typically CPU-intensive, so a high compression speed minimizes CPU usage and associated costs. High compression speed is good. The time to save a column file determines the upper bound of data ingestion. The faster we can save a file, the more a kdb+ system can ingest. In the [kdb+ tick](../architecture/tickq.md) system, the RDB is unavailable for queries during write, meaning that write speed also affects system availability.

Reviewer: There is a separate page for encryption so I wouldn't mention it here.

Author: Agreed, it is fixed now.

   1. **Decompression and encryption speed**: This reflects the time taken to restore the original file from the compressed (encrypted) version. High decompression speed means faster queries.

> There is no single best compression algorithm that outperforms all others in all aspects. You need to select compression (or avoid compression) based on your priorities. Do you need the fastest possible queries or do you minimize the storage costs? Maybe a high volume of incoming data is hitting your kdb+ system and you would like to make sure that your intraday write process can cope with the volume. Maybe you are looking for a general solution that neither excels nor falls behind in any aspect.

Reviewer 2: ```suggestion
There is no single best compression algorithm that outperforms all others in all aspects. You need to select compression (or avoid compression) based on your priorities. For example:
- Is achieving the fastest possible query execution more important to you, or do you prefer to minimize storage costs?
- Does your kdb+ system handle a high volume of incoming data, requiring a reliable intraday write process to manage the data effectively?
- Are you looking for a general solution that provides balanced performance across various aspects without excelling or underperforming in any particular area?
```

> A single thread with full use of a core can decompress approx 300MB/s, depending on data/algorithm and level.

Reviewer: Do you have a more up-to-date figure than 300MB/s from all your testing?

Author: I will publish some figures in the next PR.

---|---|---
libz.so.1 | libz.dylib<br>(pre-installed) | zlibwapi.dll<br>(32-bit and 64-bit versions available from [WinImage](http://www.winimage.com/zLibDll/index.html "winimage.com"))

> Gzip has very good compression ratio and average compression/decompression speed. Avoid high compression levels (like 8 and 9) if write speed is important for you. Gzip with level 5 is probably a good general solution.

Reviewer: ```suggestion
Gzip has very good compression ratio and average compression/decompression speed. Avoid high compression levels (like 8 and 9) if write speed is important for you. Gzip with level 5 is a good general solution.
```

---|---|---
libsnappy.so.1 | libsnappy.dylib<br>(available via package managers such as [Homebrew](https://brew.sh/) or [MacPorts](https://www.macports.org/)) | snappy.dll

> Snappy has excellent compression and decompression speed so it is a good choice if you optimize for query speed and ingestion times. Snappy falls behing the other compression solutions in compression ratio.

Reviewer: ```suggestion
Snappy has excellent compression and decompression speed so it is a good choice if you optimize for query speed and ingestion times. Snappy falls behind the other compression solutions in compression ratio.
```

    kdb+ requires at least `lz4-r129`.
    `lz4-1.8.3` works.
    We recommend using the latest `lz4` [release](https://github.com/lz4/lz4/releases) available.
    

> LZ4 is great at decompression speed and compression ratio but does not perform well in compression speed. Compression level 5 is probably a good choice if you aim fast queries and low storage costs. Avoid high compression levels (above 11).

Reviewer: ```suggestion
LZ4 is great at decompression speed and compression ratio but does not perform well in compression speed. Compression level 5 is a good choice if you aim fast queries and low storage costs. Avoid high compression levels (above 11).
```
